torch/_inductor/codecache.py

```diff
@@ -146,11 +147,83 @@ def is_gcc():
     return re.search(r"(gcc|g\+\+)", cpp_compiler())
 
 
+class _SupportedVecIsa(enum.Enum):
+    AVX512 = 1
+    AVX2 = 2
+    INVALID = -1
+
+    @staticmethod
+    def isa_str(supported_isa: enum.Enum):
+        if supported_isa == _SupportedVecIsa.AVX512:
+            return "avx512"
+        elif supported_isa == _SupportedVecIsa.AVX2:
+            return "avx2"
+        else:
+            return ""
+
+    @staticmethod
+    def vec_macro(supported_isa: enum.Enum):
+        if supported_isa == _SupportedVecIsa.AVX512:
+            return "CPU_CAPABILITY_AVX512"
+        elif supported_isa == _SupportedVecIsa.AVX2:
+            return "CPU_CAPABILITY_AVX2"
+        else:
+            return ""
+
+
+# Cache the cpuinfo to avoid I/O overhead. Meanwhile, the cpuinfo content
+# might have too much redundant content that is useless for ISA check. Hence,
+# we only cache some key isa information.
+@functools.lru_cache(1)
+def get_cpu_proc_info():
+    if sys.platform != "linux":
+        return ""
+
+    isa_info = ""
+    with open("/proc/cpuinfo") as _cpu_info:
+        _cpu_info_content = _cpu_info.read()
+        if _SupportedVecIsa.isa_str(_SupportedVecIsa.AVX512) in _cpu_info_content:
+            isa_info += _SupportedVecIsa.isa_str(_SupportedVecIsa.AVX512)
+
+        if _SupportedVecIsa.isa_str(_SupportedVecIsa.AVX2) in _cpu_info_content:
+            isa_info += _SupportedVecIsa.isa_str(_SupportedVecIsa.AVX2)
+
+        return isa_info
+
+
+def supported_vector_isa():
+    # TODO: Add ARM Vec here.
+    # Dict(k: isa, v: number of float element)
+    vec_isa_info = {
+        _SupportedVecIsa.AVX512: 16,
+        _SupportedVecIsa.AVX2: 8,
+    }
+
+    # TODO: Add windows support
+    if sys.platform != "linux":
+        return _SupportedVecIsa.INVALID
+
+    if config.cpp.simdlen is None or config.cpp.simdlen <= 1:
+        return _SupportedVecIsa.INVALID
+
+    cpu_info_content = get_cpu_proc_info()
+    for isa in vec_isa_info.keys():
+        isa_str = _SupportedVecIsa.isa_str(isa)
+        if isa_str in cpu_info_content and config.cpp.simdlen == vec_isa_info[isa]:
+            return isa
+
+    return _SupportedVecIsa.INVALID
```

Comment:
I feel the logic of handling these isa query and vector length is over-complicated... What we want is some isa tag, e.g., "avx512", "avx2" etc. and map the tag to 1) vector length; 2) macro definition, right? Would it be simpler to implement something below?
1. A lru_cached function that returns the isa tag, "avx512", "avx2" or empty string if not supported. This can be implemented as below (note that lru_cache(None) is optimal for functions without args)
```python
@lru_cache(None)
def is_avx512_supported():
    if sys.platform != 'linux':
        return False
    with open("/proc/cpuinfo", encoding="ascii") as f:
        lines = f.read()
    if "avx512" in lines:
        return "avx512"
    elif "avx2" in lines:
        return "avx2"
    else:
        return ""
```
2. Define a dict that maps the tag to a tuple (maybe NamedTuple is better to read) that contains 1) vec length; 2) macro.

Reply:
Acutally as Bin's comment, we don't want to use the isa string at difference places.

Reply:
You may consider to make it as an enum but still can simplify the code further as I mentioned.